solidity/dashboard/src/services/liquidity-rewards.js

```diff
@@ -63,34 +49,33 @@ export const fetchRewardBalance = async (address, LPrewardsContract) => {
   return await LPrewardsContract.methods.earned(address).call()
 }
 
-export const calculateAPY = (
-  totalSupplyInWei,
-  totalLPTokensCreatedInUniswapInWei,
-  pairSymbol
-) => {
-  const totalSupply = web3Utils.fromWei(totalSupplyInWei)
-  const totalLPTokensCreatedInUniswap = web3Utils.fromWei(
-    totalLPTokensCreatedInUniswapInWei
-  )
-  const totalSupplyInBN = new BigNumber(totalSupply)
-  const totalLPTokensCreatedInUniswapInBN = new BigNumber(
-    totalLPTokensCreatedInUniswap
-  )
-  //TODO : Add pairSymbol to constants (probably not be needed)
-  const rewardPoolPerWeekInBN = new BigNumber(
+export const calculateAPY = async (totalSupplyOfLPRewards, pairSymbol) => {
+  totalSupplyOfLPRewards = toTokenUnit(totalSupplyOfLPRewards)
+
+  const pairData = await getPairData(LIQUIDITY_REWARD_PAIRS[pairSymbol].address)
+
+  const rewardPoolPerWeek = new BigNumber(
     LIQUIDITY_REWARD_PAIRS[pairSymbol].rewardPoolPerWeek
   )
 
-  const totalLPTokensInLPRewardsInUSD = totalSupplyInBN
-    .multipliedBy(totalLiquidityAmount[pairSymbol])
-    .div(totalLPTokensCreatedInUniswapInBN)
+  const totalLPTokensInLPRewardsInUSD = totalSupplyOfLPRewards
+    .multipliedBy(pairData.reserveUSD)
+    .div(pairData.totalSupply)
 
-  const r = keepInUSDInBN
-    .multipliedBy(rewardPoolPerWeekInBN)
-    .div(totalLPTokensInLPRewardsInUSD)
+  const ethPrice = new BigNumber(pairData.reserveUSD).div(pairData.reserveETH)
 
-  // TODO: Add 52 to constants
-  const apy = r.plus(1).pow(52).minus(1)
+  let keepTokenInUSD = 0
+  if (pairData.token0.symbol === "KEEP") {
+    keepTokenInUSD = ethPrice.multipliedBy(pairData.token0.derivedETH)
+  } else if (pairData.token1.symbol === "KEEP") {
+    keepTokenInUSD = ethPrice.multipliedBy(pairData.token1.derivedETH)
+  } else {
+    keepTokenInUSD = await getKeepTokenPriceInUSD()
```

Comment:
Finish step is to remove the if statement and ethPrice variable from the `calculateAPY` function in **services/liquidity-rewards.js** file nd replace it with:

```
  const keepTokenInUSD = pairData.keepTokenPriceInUSD

  const r = new BigNumber(keepTokenInUSD)
    .multipliedBy(rewardPoolPerWeek)
    .div(totalLPTokensInLPRewardsInUSD)
```

Note that we are calling `new BigNumber(...)` in the `r` variable since we changed the type of the `keepTokenPriceInUSD` to a string.

Reply:
Last but not least we could mark the `getKeepTokenPriceInUSD` function as @private in the JSDoc comment (above the function), because we need the pair data from Uniswap anyway to get the price so we might as-well get the pair with already calculated price by calling `getPairData() `

Reply:
Thoughts? I'm always open to a discussion 😄

Reply:
```
export const getKeepTokenPriceInUSD = async (pairData) => {
  const ethPrice = new BigNumber(pairData.reserveUSD).div(pairData.reserveETH)
  if (pairData.token1.symbol === "KEEP") {
    return ethPrice.multipliedBy(pairData.token1.derivedETH)
  } else {
    return ethPrice.multipliedBy(pairData.token0.derivedETH)
  }
}
```

This is only true for a Uniswap pair with KEEP. It would not work with TBTC-ETH. We need KEEP token price in USD for every LPRewards calculation since rewards are claimed in KEEP. So if we have the KEEP token in a given pair we can calculate the KEEP token price based on fetched pair data otherwise, we need to call external API.

Reply:
> Then in `getPairData` function I would add this code to assign the keep token USD price to a property with the help of [spread syntax (...)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)
> 
> ```
>   const keepTokenPriceInUSD = await getKeepTokenPriceInUSD(
>     response.data.data.pair
>   )
> 
>   return {
>     usdPrice: keepTokenPriceInUSD.toString(),
>     ...response.data.data.pair
>   }
> ```
> 
> We are changing the usd price to a string because all other properties returned from the Uniswap api are also strings.

Same thing as above, it would only work for the Uniswap pair with KEEP. Also, I believe `getPairData` should be as generic as possible and should not include `keepTokenPriceInUSD: keepTokenPriceInUSD.toString()` because it is not correct for the other pairs.

Reply:
Yes, i see it now - my bad 😄  You are 100% correct. Let's KEEP it at it is 🔥